ibm/resource_ibm_network_gateway.go

```diff
@@ -0,0 +1,857 @@
+			"networkGatewayMember[networkGatewayId]," +
+			"memoryCapacity,powerSupplyCount,",
+	).GetObject()
+
+	if err != nil {
+		return fmt.Errorf("Error retrieving Network Gateway: %s", err)
+	}
+
+	d.Set("hostname", *result.Hostname)
+	d.Set("domain", *result.Domain)
+	d.Set("networkGatewayId", *result.NetworkGatewayMember.NetworkGatewayId)
+
+	if result.Datacenter != nil {
+		d.Set("datacenter", *result.Datacenter.Name)
+	}
+
+	d.Set("network_speed", *result.PrimaryNetworkComponent.MaxSpeed)
+	if result.PrimaryIpAddress != nil {
+		d.Set("public_ipv4_address", *result.PrimaryIpAddress)
+	}
+	d.Set("private_ipv4_address", *result.PrimaryBackendIpAddress)
+
+	d.Set("private_network_only", *result.PrivateNetworkOnlyFlag)
+
+	if result.PrimaryNetworkComponent.NetworkVlan != nil {
+		d.Set("public_vlan_id", *result.PrimaryNetworkComponent.NetworkVlan.Id)
+	}
+
+	if result.PrimaryBackendNetworkComponent.NetworkVlan != nil {
+		d.Set("private_vlan_id", *result.PrimaryBackendNetworkComponent.NetworkVlan.Id)
+	}
+
+	d.Set("notes", sl.Get(result.Notes, nil))
+	d.Set("memory", *result.MemoryCapacity)
+
+	d.Set("redundant_network", false)
+	d.Set("unbonded_network", false)
+
+	backendNetworkComponent, err := service.Filter(
+		filter.Build(
+			filter.Path("backendNetworkComponents.status").Eq("ACTIVE"),
+		),
+	).Id(id).GetBackendNetworkComponents()
+
+	if err != nil {
+		return fmt.Errorf("Error retrieving Network Gateway network: %s", err)
+	}
+
+	if len(backendNetworkComponent) > 2 && result.PrimaryBackendNetworkComponent != nil {
+		if *result.PrimaryBackendNetworkComponent.RedundancyEnabledFlag {
+			d.Set("redundant_network", true)
+		} else {
+			d.Set("unbonded_network", true)
+		}
+	}
+
+	tagReferences := result.TagReferences
+	tagReferencesLen := len(tagReferences)
+	if tagReferencesLen > 0 {
+		tags := make([]string, 0, tagReferencesLen)
+		for _, tagRef := range tagReferences {
+			tags = append(tags, *tagRef.Tag.Name)
+		}
+		d.Set("tags", tags)
+	}
+
+	connInfo := map[string]string{"type": "ssh"}
+	if !*result.PrivateNetworkOnlyFlag && result.PrimaryIpAddress != nil {
+		connInfo["host"] = *result.PrimaryIpAddress
+	} else {
+		connInfo["host"] = *result.PrimaryBackendIpAddress
+	}
+	d.SetConnInfo(connInfo)
+	d.Set("associated_vlans", resourceIBMNetworkGatewayVlanAssociatedReader(d, meta))
+
+	return nil
+}
+
+func resourceIBMNetworkGatewayUpdate(d *schema.ResourceData, meta interface{}) error {
+	id, _ := strconv.Atoi(d.Id())
+	if d.HasChange("tags") {
+		err := setHardwareTags(id, d, meta)
+		if err != nil {
+			return err
+		}
+	}
+
+	if d.HasChange("notes") {
+		err := setHardwareNotes(id, d, meta)
+		if err != nil {
+			return err
+		}
+	}
+
+	if d.HasChange("associated_vlans") {
+		if v, ok := d.GetOk("associated_vlans"); ok && v.(*schema.Set).Len() > 0 {
+			associatedVlans := expandVlans(v.(*schema.Set).List())
+			resourceIBMNetworkGatewayVlanAssociate(d, meta, associatedVlans, id)
```

Comment:
If some vlans are deleted in the config, then those must be deleted via https://sldn.softlayer.com/reference/services/SoftLayer_Network_Gateway_Vlan/deleteObjects.